crates/bevy_ecs/macros/src/component.rs

```diff
@@ -8,20 +8,76 @@ pub fn derive_resource(input: TokenStream) -> TokenStream {
     let mut ast = parse_macro_input!(input as DeriveInput);
     let bevy_ecs_path: Path = crate::bevy_ecs_path();
 
-    ast.generics
-        .make_where_clause()
+    let attrs = match parse_resource_attr(&ast) {
+        Ok(attrs) => attrs,
+        Err(e) => return e.into_compile_error().into(),
+    };
+
+    let is_sync = attrs.is_sync;
+
+    let where_clause = ast.generics.make_where_clause();
+    where_clause
         .predicates
-        .push(parse_quote! { Self: Send + Sync + 'static });
+        .push(parse_quote! { Self: Send + 'static });
+
+    // Add a bound to ensure that the `Sync`-ness is correct.
+    if is_sync {
+        where_clause.predicates.push(parse_quote! { Self: Sync });
+    }
 
     let struct_name = &ast.ident;
     let (impl_generics, type_generics, where_clause) = &ast.generics.split_for_impl();
 
     TokenStream::from(quote! {
-        impl #impl_generics #bevy_ecs_path::system::Resource for #struct_name #type_generics #where_clause {
+        // SAFETY: The trait bounds on this impl guarantee that the non-`Sync` types will not have `IS_SYNC = true`.
+        unsafe impl #impl_generics #bevy_ecs_path::system::Resource for #struct_name #type_generics #where_clause {
+            const IS_SYNC: bool = #is_sync;
```

Comment:
These should definitely have some compile fail tests to ensure that this breaks when this contract is broken.

Separately, we should see if the compile failures are readable by normal users.

Reply:
I added a compile fail test. The error message isn't great, but it isn't any worse than before this PR.

I tried a few ways of providing a more descriptive error message, but they all caused problems in edge cases.

Reply:
After investigating further, the error message is actually good for concrete types. It shows up when you define the struct, and points you directly to the field that makes the struct `!Sync`.

It's less good for generic types, because the error doesn't occur until you try to use the resource.

```rust
// Concrete types:

#[derive(Resource)] // error message shows up here, right where we want it.
struct MyResource(Cell<u32>);
```

```rust
// Generic types:

#[derive(Resource)]
struct MyResource<T>(Cell<T>);

// The error message doesn't show up until here, which is un-ideal.
fn my_system(_: Res<MyResource<u32>>) {}
```

IMO this isn't perfect, but it's perfectly acceptable.